### PR TITLE
Extract generate_index command to rubygems-generate_index gem

### PR DESCRIPTION
**Pull request by segiddins**
So generate_index can be implemented with dependencies, such as the compact index

Took this approach from feedback in https://github.com/rubygems/rubygems/pull/6853

Running `gem generate_index` by default will use an installed rubygems-generate_index, or install and then use the command from the gem

New gem extraction in https://github.com/rubygems/rubygems-generate_index/pull/2

### CODE REVIEW

File: bundler/spec/support/builders.rb [hunk 1]
```diff
@@ -258,7 +258,12 @@ def update_repo(path)
       @_build_repo = File.basename(path)
       yield
       with_gem_path_as Path.base_system_gem_path do
-        gem_command :generate_index, :dir => path
+        generate_index_path = Dir[Spec::Path.base_system_gem_path.join("gems/rubygems-generate_index*/lib")].first ||
```
Comment (segiddins) [inline bundler/spec/support/builders.rb:261]: remove `||`

File: tool/bundler/test_gems.rb [hunk 2]
```diff
@@ -5,10 +5,11 @@
...
 gem "rack-test", "~> 1.1"
-gem "compact_index", "~> 0.13.0"
+gem "compact_index", "~> 0.14.0"
 gem "sinatra", "~> 2.0"
 # for Ruby 2.6. bundler/spec/install/gemfile/sources_spec.rb is failed with sinatra-2.0.8.1 and tilt-2.2.0.
 gem "tilt", "~> 2.0.11"
 gem "rake", "13.0.1"
 gem "builder", "~> 3.2"
 gem "rb_sys"
+gem "rubygems-generate_index", path: "/Users/segiddins/Development/github.com/rubygems/rubygems-generate_index"
```
Comment (segiddins) [inline tool/bundler/test_gems.rb:15]: use released gem

**Review by indirect — APPROVED** [review]

File: lib/rubygems/commands/generate_index_command.rb [hunk 3]
```diff
@@ -1,86 +1,51 @@
...
-               "Generate indexes for RubyGems",
-               "(always true)" do |value, options|
-      options[:build_modern] = value
+      def invoke_with_build_args(args, build_args)
+        name = "rubygems-generate_index"
+        spec = begin
+          Gem::Specification.find_by_name(name)
+        rescue Gem::LoadError
+          require "rubygems/dependency_installer"
+          Gem.install(name, Gem::Requirement.default, Gem::DependencyInstaller::DEFAULT_OPTIONS).find {|s| s.name == name }
```
Comment (simi) [inline lib/rubygems/commands/generate_index_command.rb:24]: Any idea which gem source is used in here?

Comment (segiddins) [reply lib/rubygems/commands/generate_index_command.rb:24]: I believe it is the default ?

File: lib/rubygems/commands/generate_index_command.rb [hunk 4]
```diff
@@ -1,86 +1,51 @@
...
-  def initialize
-    super "generate_index",
-          "Generates the index files for a gem server directory",
-          :directory => ".", :build_modern => true
+unless defined? Gem::Commands::GenerateIndexCommand
+  class Gem::Commands::GenerateIndexCommand < Gem::Command
+    module RubygemsTrampoline
+      def description # :nodoc:
+        <<~EOF
+          The server command has been moved to the rubygems-server gem.
```
Comment (hsbt) [inline lib/rubygems/commands/generate_index_command.rb:10]:
```suggestion
          The generate_index command has been moved to the rubygems-generate_index gem.
```

File: lib/rubygems/commands/generate_index_command.rb [hunk 5]
```diff
@@ -1,86 +1,51 @@
...
+        EOF
+      end
 
-    add_option "-d", "--directory=DIRNAME",
-               "repository base dir containing gems subdir" do |dir, options|
-      options[:directory] = File.expand_path dir
-    end
+      def execute
+        alert_error "Install the rubygems-server gem for the server command"
```
Comment (hsbt) [inline lib/rubygems/commands/generate_index_command.rb:15]:
```suggestion
        alert_error "Install the rubygems-generate_index gem for the generate_index command"
```

File: lib/rubygems/commands/generate_index_command.rb [hunk 6]
```diff
@@ -1,86 +1,51 @@
...
-    deprecate_option("--no-modern", version: "4.0", extra_msg: "The `--no-modern` option is currently ignored. Modern indexes (specs, latest_specs, and prerelease_specs) are always generated.")
-
-    add_option "--update",
-               "Update modern indexes with gems added",
-               "since the last update" do |value, options|
-      options[:update] = value
+    # remove_method(:initialize) warns, but removing new does not warn
+    def self.new
+      command = allocate
+      command.send(:initialize, "server", "Starts up a web server that hosts the RDoc (requires rubygems-server)")
```
Comment (hsbt) [inline lib/rubygems/commands/generate_index_command.rb:45]:
```suggestion
      command.send(:initialize, "generate_index", "Generates the index files for a gem server directory (requires rubygems-generate_index)")
```

**Review by hsbt — COMMENTED** [review]
I'm not sure these are intentional. Can you confirm them?

**Comment by segiddins** [comment]
Ah thanks for catching those. I was doing my testing with rubygems-server and copy/pasted

**Comment by hsbt** [comment]
Bundler spec with `ruby-core` is failing now.

https://github.com/rubygems/rubygems/actions/runs/7053171654/job/19315609030?pr=7085

```
        ERROR:  While executing gem ... (ArgumentError)
            comparison of CompactIndex::GemVersion with CompactIndex::GemVersion failed
        	/home/runner/work/rubygems/rubygems/ruby/ruby/tmp/4/gems/base/ruby/3.3.0+0/gems/rubygems-generate_index-1.1.0/lib/rubygems/indexer.rb:583:in `sort!'
```

I could reproduce it with my local box. I'm not sure why only failed with ruby-core test suite.

**Review by hsbt — APPROVED** [review]
Thanks @segiddins ! I also confirmed to fix with `rubygems-generate_index` 1.1.2.

**Comment by voxik** [comment]
Just FTR, this change breaks gem rubygems-mirror as reported [here](https://lists.fedoraproject.org/archives/list/ruby-sig@lists.fedoraproject.org/message/V7M4LAQH5DU6CBEANNT5NGEL77HF4Q3V/):

~~~
$ ruby -Ilib -e 'Dir.glob "./test/test_*.rb", &method(:require)'
<internal:/usr/share/rubygems/rubygems/core_ext/kernel_require.rb>:127:in `require': cannot load such file -- rubygems/indexer (LoadError)
    from <internal:/usr/share/rubygems/rubygems/core_ext/kernel_require.rb>:127:in `require'
    from /builddir/build/BUILD/rubygems-mirror-1.3.0/usr/share/gems/gems/rubygems-mirror-1.3.0/lib/rubygems/mirror/test_setup.rb:7:in `<top (required)>'
    from <internal:/usr/share/rubygems/rubygems/core_ext/kernel_require.rb>:127:in `require'
    from <internal:/usr/share/rubygems/rubygems/core_ext/kernel_require.rb>:127:in `require'
    from /builddir/build/BUILD/rubygems-mirror-1.3.0/usr/share/gems/gems/rubygems-mirror-1.3.0/test/test_gem_mirror.rb:3:in `<top (required)>'
    from <internal:/usr/share/rubygems/rubygems/core_ext/kernel_require.rb>:127:in `require'
    from <internal:/usr/share/rubygems/rubygems/core_ext/kernel_require.rb>:127:in `require'
    from <internal:dir>:411:in `glob'
    from -e:1:in `<main>'
~~~

Not that I disagree with the change or I care that much about rubygems-mirror. But unfortunately, this is for the [second](https://github.com/rubygems/rubygems/pull/6623) [time](https://github.com/enkessler/childprocess/issues/190) in short period of time, where something was removed from RubyGems what breaks other projects.

Therefore, could stuff like this be deprecated for a while?

**Pull request merged**